src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1316,3 +1322,89 @@ uct_dc_mlx5_query_tl_devices(uct_md_h md, uct_tl_device_resource_t **tl_devices_
               uct_dc_mlx5_iface_config_t);
+
+static void
+uct_dc_mlx5_dci_keepalive_handle_failure(uct_dc_mlx5_iface_t *iface,
+                                         struct mlx5_cqe64 *cqe,
+                                         uint8_t dci,
+                                         ucs_status_t ep_status)
+{
+    uct_ib_iface_t *ib_iface = &iface->super.super.super;
+    uct_dc_mlx5_ep_t *ep;
+    uct_rc_txqp_t *txqp;
+    uct_ib_mlx5_txwq_t *txwq;
+    uct_rc_iface_send_op_t *op;
+    uint16_t hw_ci;
+    ucs_queue_elem_t *elem;
+    ucs_status_t status;
+    ucs_log_level_t log_lvl;
+
+    ucs_assert(dci == iface->tx.ndci);
+
+    hw_ci  = ntohs(cqe->wqe_counter);
+    txqp   = &iface->tx.dcis[dci].txqp;
+    elem   = ucs_queue_pull(&txqp->outstanding);
+    if (elem == NULL) {
+        /* outstanding list is empty, just exit */
+        goto reset_dci;
+    }
+
+    op = ucs_container_of(elem, uct_rc_iface_send_op_t, queue);
+    if (hw_ci != op->sn) {
+        goto put_op;
+    }
+
+    ep = ucs_derived_of(op->ep, uct_dc_mlx5_ep_t);
+    if (ep->dci != UCT_DC_MLX5_EP_NO_DCI) {
+        uct_dc_mlx5_ep_handle_failure(ep, cqe, ep_status);
+        goto put_op;
+    }
+
+    ucs_assert(ep != iface->tx.fc_ep);
+    status  = ib_iface->ops->set_ep_failed(ib_iface, &ep->super.super,
+                                           ep_status);
+    log_lvl = uct_ib_iface_failure_log_level(ib_iface, status, ep_status);
+
+    if (ep_status != UCS_ERR_CANCELED) {
+        uct_ib_mlx5_completion_with_err(ib_iface, (void*)cqe,
+                                        &iface->tx.dcis[dci].txwq, log_lvl);
+    }
+
+put_op:
+    ucs_mpool_put(op);
+
+reset_dci:
+    txwq        = &iface->tx.dcis[dci].txwq;
+    uct_rc_txqp_available_set(txqp, (int16_t)iface->super.super.config.tx_qp_len);
+    /* TODO: re-schedule other outstanding */
+    uct_rc_txqp_purge_outstanding(&iface->super.super, txqp, ep_status,
+                                  txwq->sw_pi, 0);
+    status = uct_dc_mlx5_iface_reset_dci(iface, &iface->tx.dcis[dci]);
+    if (status != UCS_OK) {
+        ucs_fatal("iface %p failed to reset dci[%d] qpn 0x%x: %s",
+                  iface, dci, txwq->super.qp_num, ucs_status_string(status));
+    }
+
+    status = uct_dc_mlx5_iface_dci_connect(iface, &iface->tx.dcis[dci]);
+    if (status != UCS_OK) {
+        ucs_fatal("iface %p failed to connect dci[%d] qpn 0x%x: %s",
+                  iface, dci, txwq->super.qp_num, ucs_status_string(status));
+    }
+}
+
+static uct_dc_dci_ops_t uct_dc_mlx5_dci_keepalive_ops = {
+    .handle_failure = uct_dc_mlx5_dci_keepalive_handle_failure
+};
+
+ucs_status_t uct_dc_mlx5_iface_keepalive_init(uct_dc_mlx5_iface_t *iface)
+{
+    if (ucs_likely(iface->tx.keepalive_ndci)) {
+        return UCS_OK;
+    }
+
+    ucs_assert(iface->tx.ndci <= UCT_DC_MLX5_IFACE_MAX_USER_DCIS);
+
+    iface->tx.keepalive_ndci = 1;
+    return uct_dc_mlx5_iface_create_dci(iface, &uct_dc_mlx5_dci_keepalive_ops,
```

Comment:
use some constant instead of 1?
or if it's always 1 DCI, replace keepalive_ndci by some flag on the iface